libsolidity/interface/CompilerStack.cpp

```diff
@@ -399,27 +428,44 @@ bool CompilerStack::analyze()
 			if (source->ast && !typeChecker.checkTypeRequirements(*source->ast))
 				noErrors = false;
 
+		// Create & assign callgraphs and check for contract dependency cycles
 		if (noErrors)
 		{
+			// Cycles we found, used to avoid duplicate reports for the same dependency
+			map<ContractDefinition const*, ASTNode const*, ASTNode::CompareByID> foundCycles;
+			util::CycleDetector<ContractDefinition> cycleDetector{[&](ContractDefinition const& _contract, util::CycleDetector<ContractDefinition>& _cycleDetector, size_t _depth)
+			{
+				if (_depth >= MaxCycleDetectionRecursionDepth)
+					m_errorReporter.fatalTypeError(7864_error, _contract.location(), "Contract dependencies exhausting cyclic dependency validator");
+
+				for (auto& [dependencyContract, referencee]: _contract.annotation().contractDependencies)
+					if (_cycleDetector.run(*dependencyContract))
+						return;
+			}};
+
 			for (Source const* source: m_sourceOrder)
 				if (source->ast)
 					for (ASTPointer<ASTNode> const& node: source->ast->nodes())
 						if (auto const* contractDefinition = dynamic_cast<ContractDefinition*>(node.get()))
 						{
-							Contract& contractState = m_contracts.at(contractDefinition->fullyQualifiedName());
-
-							contractState.contract->annotation().creationCallGraph = make_unique<CallGraph>(
-								FunctionCallGraphBuilder::buildCreationGraph(
-									*contractDefinition
-								)
-							);
-							contractState.contract->annotation().deployedCallGraph = make_unique<CallGraph>(
-								FunctionCallGraphBuilder::buildDeployedGraph(
-									*contractDefinition,
-									**contractState.contract->annotation().creationCallGraph
-								)
-							);
+							createAndAssignCallGraphs(*contractDefinition);
+							ContractDefinition const* cycleContract = cycleDetector.run(*contractDefinition);
+							if (cycleContract)
+								foundCycles.emplace(cycleContract, contractDefinition->annotation().contractDependencies[cycleContract]);
 						}
+
+			for (auto const& [cycle, referencee]: foundCycles)
+			{
+				SecondarySourceLocation secondaryLocation{};
+				secondaryLocation.append("Referenced contract"s, cycle->location());
+
+				m_errorReporter.typeError(
+					7813_error,
+					referencee->location(),
+					secondaryLocation,
+					"Circular reference found."
+				);
+			}
```

Comment:
I'd extract more of this code into separate functions to keep `analyze()` leaner.